### PR TITLE
Add ECMP CID algorithm

### PR DESCRIPTION
**Pull request by ianswett**
I tried to make this a subset of the plaintext algorithm, but the differences were substantial enough(ie: there are no server IDs) that I created a new algorithm.

### CODE REVIEW

File: draft-ietf-quic-load-balancers.md [hunk 1]
```diff
@@ -359,6 +359,47 @@ in consecutive octets beginning with the second. All other bits in the
...
+
+### Configuration Agent Actions
+
+The configuration agent selects a length for the server ID encoding. This
+length MUST have enough entropy to have a different code point for each server.
+
+### Load Balancer Actions
+
+On each incoming packet, the load balancer extracts consecutive octets,
+beginning with the second octet. If these octets match and existing server ID,
```
Comment (nibanks) [inline draft-ietf-quic-load-balancers.md:389]:
```suggestion
beginning with the second octet. If these octets match an existing server ID,
```

**Comment by nibanks** [comment]
@ianswett I'm a bit confused on the difference here with the vanilla plaintext algorithm. Is the only difference that the LB routes initial packets specially? Does the existing plaintext algorithm preclude that behavior?

**Comment by ianswett** [comment]
I don't think it's precluded, but I do think it's different enough it's worth documenting, but I realize the difference is subtle.  @martinduke had the same thought, so I'll try to further clarify this PR.

This is a way to adapt existing stateful flow-tracking ECMP L4 load-balancers to QUIC with almost no server coordination.  The plaintext requires substantially more coordination and requires leaving some areas of the connection ID space unusable in order to allow for adding servers.  When each new flow comes in, the L4 load-balancer can decide where to send it, increasing the evenness of traffic.

**Comment by martinduke** [comment]
Thanks for the PR. There are definite simplicity benefits to not having explicitly assigned server IDs; if we can work through the issues here, I'd just as soon use this framework to *replace* PCID.

Presumably the LB is maintaining a table of its SID assignments, and the servers are maintaining lists of all SIDs they've observed.

Also, if a server goes down, it seems straightforward that the LB would delete its entry and those SIDs would be free to reallocation to other servers.

We would rewrite the non-compliant DCID section a bit, but that seems simple enough.

So here are some issues:
- If the clients always initiate with a DCID that is shorter than the minimum, then there is no way for the server to obtain its first server ID. So there would appear to be a strict limit of 7 byte SIDs as v1 specifies a minimum CIDL of 8.
- I think the SID space has to be pretty big because this approach will be profligate with SIDs. Most new connections will consume a new SID codepoint for the existing pool of servers. This is going to be a pretty memory-intensive table! You will probably needs hundreds of thousands (?) of times more entries than you would have for PCID.
- Profligate assignment of SIDs will eventually crowd out the space to bring any new servers on unless other servers die and release their SIDs. Unless there's some kind of timeout mechanism? [1] [2]

[1] This mechanism could be something like, "If a server hasn't received a packet in X minutes for a certain SID, it MUST retire that CID with that SID on any connection still in use" coupled with a load balancer timeout on its mapping after X minutes.
[2] Alternatively, we could do something clever with config rotation where, when the table gets full, or there's a new server, the LB wakes up the config agent to give everyone new CR bits, and this is a cue to flush the SID table.

**Comment by martinduke** [comment]
Filed #64 as a tracking issue.

**Comment by nibanks** [comment]
I'm still not convinced we should have a separate algorithm for this. Instead, I'd argue we should ensure the existing plaintext algorithm is generic enough to handle all related scenarios.

**Comment by martinduke** [comment]
@ianswett and discussed this proposal. Google's operational experience is that the memory needs are not onerous. We agreed that this algorithm should have a "LB Idle timeout" parameter after which SID allocations will be freed. Also, server IDs must be no more 7 octets in length (or else 8B client-generated CIDs will not successfully populate the server's table).

The advantages of this with respect to PCID is that there is much less configuration and that it conforms more closely to some existing load balancers. The disadvantage is that it requires much more per-connection state at the load balancer.

**Comment by martinduke** [comment]
I am coming around on this design. When this PR is revised, we will go to the list with three options:

1) Have two unencrypted algorithms
2) Embrace low-state, high-configuration and stick with existing PCID
3) Pick the opposite and replace PCID with this design

**Comment by martinduke** [comment]
As @ianswett is inundated with IESG comments right now, I'm going to commit this PR, make some other edits along the lines of what we agreed, and then take the larger issue to the list.

**Pull request merged**